lib/core/Resolver.ts

```diff
@@ -2,79 +2,231 @@ import AnchoredOperationModel from './models/AnchoredOperationModel';
+        updateOperations.push(operation);
+      } else {
+        // This is a deactivate operation.
+        deactivateOperations.push(operation);
+      }
+    }
+    return {
+      createOperations,
+      recoverOperations,
+      updateOperations,
+      deactivateOperations
+    };
+  }
+
+  /**
+   * Iterate through all duplicates of creates until we can construct an initial DID state (some creates maybe incomplete. eg. without `delta`).
+   */
+  private async applyCreateOperation (createOperations: AnchoredOperationModel[]): Promise<DidState | undefined> {
+    let didState;
+
+    for (const createOperation of createOperations) {
+      didState = await this.applyOperation(createOperation, undefined);
+
+      // Exit loop as soon as we can construct an initial state.
+      if (didState !== undefined) {
+        break;
+      }
+    }
+
+    return didState;
+  }
+
+  /**
+   * Apply recovery/deactivate operations until an opertaion matching the next recovery commitment cannot be found.
+   */
+  private async applyRecoverAndDeactivateOperations (startingDidState: DidState, commitValueToOperationMap: Map<string, AnchoredOperationModel[]>)
+    : Promise<DidState> {
+    let didState = startingDidState;
+
+    while (commitValueToOperationMap.has(didState.nextRecoveryCommitmentHash!)) {
+      let operationsWithCorrectRevealValue: AnchoredOperationModel[] = commitValueToOperationMap.get(didState.nextRecoveryCommitmentHash!)!;
+
+      // Sort using blockchain time.
+      operationsWithCorrectRevealValue = operationsWithCorrectRevealValue.sort((a, b) => a.transactionNumber - b.transactionNumber);
+
+      const newDidState: DidState | undefined = await this.applyFirstValidOperation(operationsWithCorrectRevealValue, didState);
+
+      // We are done if we can't find a valid recover/deactivate operation to apply.
+      if (newDidState === undefined) {
+        break;
+      }
+
+      // We reach here if we have successfully computed a new DID state.
+      didState = newDidState;
+
+      // If the previous applied operation is a deactivate. No need to continue further.
+      if (didState.nextRecoveryCommitmentHash === undefined) {
+        return didState;
+      }
+    }
+
+    return didState;
+  }
+
+  /**
+   * Apply update operations until an opertaion matching the next update commitment cannot be found.
+   */
+  private async applyUpdateOperations (startingDidState: DidState, commitValueToOperationMap: Map<string, AnchoredOperationModel[]>)
+    : Promise<DidState> {
+    let didState = startingDidState;
 
-    // Apply "update/delta" operations.
-    didState = await this.applyOperations(updateOperationsToBeApplied, didState);
+    while (commitValueToOperationMap.has(didState.nextUpdateCommitmentHash!)) {
+      let operationsWithCorrectRevealValue: AnchoredOperationModel[] = commitValueToOperationMap.get(didState.nextUpdateCommitmentHash!)!;
+
+      // Sort using blockchain time.
```

Comment:
is this list size controllable by an attacker? In other words, once a reveal value is published, can an attacker replay it to cause this list to be arbitrarily large?

Reply:
An attacker can always _attempt_ to replay any commit values, even invalid ones. However, the entire point of the commit reveal scheme is that the hash check to determine validity is much faster than a sig check, and because we're ledger time ordering for each claimed commitment, replayed commits will usually never be evaluated, because the user's correct commit will be encountered first and used. Because the user's commitments generally will appear first, their successful eval mean you can simply stop iterating that commit's bucket of possible matching ops.

Reply:
Orie has a good point about the list being too large though but I'm not sure if we have a good way to block it.

Reply:
One cannot prevent an attacker playing an op using the same reveal value, the simplest way to do so is just keep replaying the same valid op again and again. But as @csuwildcat pointed out, we will process the the first valid one in blockchain time and discard the rest.

One thing worth noting is that in operation store we already index by blockchain time and we get by sorting in this order. So this sort here is potentially "not needed", and likely an order N operation in practice.